[16.0][ADD] hr_timesheet_predefined_description_rules

Anxo82 opened the pull request:
@ValentinVinagre @HaraldPanten 

[T-8695]

Anxo82 committed:
[ADD] hr_timesheet_predefined_description_rules

luis-ron reviewed (changes requested):
@Anxo82 I have a few requests:

1. Following the path: Timesheets -> My Timesheets -> My Timesheet Sheets

<img width="686" height="232" alt="image" src="https://github.com/user-attachments/assets/f73ad6ee-0f3c-4c91-a6ca-8ebe6a01b1a9" />

When I select the “New” option, I don't see the option to use what I have previously configured.

<img width="1156" height="407" alt="image" src="https://github.com/user-attachments/assets/306596f9-f49c-4172-8446-f68c644520d1" />

2. I have noticed that by following the path Timesheets -> My Timesheets -> Timesheets to submit. 
When I create a new one, you can set the default values in the Predefined Description field, but they do not behave correctly.

<img width="1612" height="550" alt="image" src="https://github.com/user-attachments/assets/2726b81a-f564-4ced-a8cd-34bf810b681d" />

As you can see in the last image, the field Description is not autocompleted by the rules i have configured.

3. On the other hand, following the same model, I have noticed that if you modify this record and enter the same predefined value, when you save it, the description value is overwritten.

I am attaching a short video so you can see it (Demo_1).
[Demo_1.webm](https://github.com/user-attachments/assets/6d200b03-750e-4d29-9241-4b1b0f0282ce)

Thanks in advance!

Anxo82 commented:
@luis-ron Added `predefined_description_id` to _Timesheets > My Timesheets Sheets > New > Details_ :
<img width="2534" height="544" alt="image" src="https://github.com/user-attachments/assets/d919a0a9-59b3-4c30-b132-af88161318e4" />
Also added to Timesheets > My Timesheets Sheets > New > Details
<img width="2534" height="728" alt="image" src="https://github.com/user-attachments/assets/4c26cd88-7196-416c-8343-2a1ec3eca869" />
On the other hand, in _My Timesheets Submit > New_ it doesn’t work because there is no `task_id`
<img width="2534" height="728" alt="image" src="https://github.com/user-attachments/assets/61349086-4ece-48c4-a3d0-3dd477355fbb" />
Added dependency hr_timesheet_sheet to allow including the field in those menus.
Thanks

luis-ron reviewed (approved):
Thanks for the changes @Anxo82 

Functional review: LGTM 👍🏻

ValentinVinagre commented on hr_timesheet_predefined_description_rules/models/hr_timesheet_predefined_description.py:
```diff
@@ -0,0 +1,34 @@
+
+    allow_free_text = fields.Boolean(
+        help=(
+            "Enable this to allow users to enter additional text when "
```
"this"? 😄  change please jaja

ValentinVinagre commented on hr_timesheet_predefined_description_rules/models/hr_timesheet_predefined_description.py:
```diff
@@ -0,0 +1,34 @@
+        ),
+    )
+
+    @api.onchange("allow_free_text")
```
depends not working?

Anxo82 replied:
I think using `@api.onchange` is better here because it only affects the form view, and adding two separate computed fields to achieve the same effect would be redundant.

ValentinVinagre commented on hr_timesheet_predefined_description_rules/models/account_analytic_line.py:
```diff
@@ -0,0 +1,45 @@
+class AccountAnalyticLine(models.Model):
+    _inherit = "account.analytic.line"
+
+    @api.onchange("predefined_description_id", "name", "task_id")
```
onchange and not depends?

Anxo82 replied:
I think it’s better to use `@api.onchange` here because it only affects the form view, and using a `@api.depends` to do the same would be overkill and could overwrite what the user types.

ValentinVinagre commented on hr_timesheet_predefined_description_rules/models/account_analytic_line.py:
```diff
@@ -0,0 +1,45 @@
+            current = (rec.name or "").strip()
+
+            if predef.allow_free_text and predef.free_text_required:
+                if current == base or not current:
```
add it to the previous check

ValentinVinagre commented on hr_timesheet_predefined_description_rules/models/hr_timesheet_predefined_description.py:
```diff
@@ -0,0 +1,34 @@
+            "using this predefined description."
+        ),
+    )
+    free_text_required = fields.Boolean(
```
What happens if you activate this option and time records have already been created using that option? The system will start to display errors.
Can you check it?

Anxo82 replied:
I’ve checked it, and I haven’t seen any errors when activating this option on existing time records.

luis-ron reviewed (approved):
After the latest technical changes made to the module, I’m offering a new functional validation. LGTM."

ValentinVinagre commented on hr_timesheet_predefined_description_rules/models/account_analytic_line.py:
```diff
@@ -0,0 +1,51 @@
+                else False
+            )
+            predef = (
+                self.env["timesheet.predefined.description"].browse(predef_id)
```
add .exist

Anxo82 replied:
@ValentinVinagre changes done! Thanks

Anxo82 commented:
@lbarry-apsl  or @etobella  could you please review this PR for merging? Thanks!

etobella reviewed (approved):
/ocabot merge nobump
